src/KUSignalLib/DSP/DSP.py

```diff
@@ -25,4 +26,114 @@ def IIRDirectForm2(b, a, x):
         tmp = np.dot(denominator, delayLine) #df2 left side
         delayLine[1:] = delayLine[:-1] #shift delay line
         delayLine[0] = x[i]*a[0] + tmp #new value is x[n] * a[0] + sum of left side
-    return y
\ No newline at end of file
+    return y
+
+def Interpolate(x, n, mode="linear"):
+    """
+    Perform interpolation on an upsampled signal.
+
+    :param x: Input signal.
+    :param n: Upsampled factor (signal is already upsampled)
+    :param mode: Interpolation type -> "linear", "quadratic"
+    :return: Interpolated signal.
+    """
+    nonzero_indices = np.arange(0, len(x)*n, n) # Generate indices for upsampled signal
+    interpolation_function = intp.interp1d(nonzero_indices, np.array(x), kind=mode, fill_value='extrapolate') # create interpolation function
+    interpolated_signal = interpolation_function(np.arange(len(x)*n)) # interpolate the signal
+    return interpolated_signal
+
+
+def Upsample(x, n, dim=1, interpolate=True, axis=0):
+    """
+    Discrete signal upsample implementation.
+
+    :param x: Input signal.
+    :param n: Upsample factor.
+    :param dim: Dimension of input signal. Default is 1.
+    :param interpolate: Flag indicating whether to perform interpolation. Default is True.
+    :param axis: Axis of x that is upsampled. Default is 0.
+    :return: Upsampled signal.
+    """
+    x_upsampled = []  # Initialize a list to store the upsampled signal
+
+    if dim == 1:  # If the input signal is 1D
+        if interpolate:
+            x_upsampled = Interpolate(x, n, mode="linear")
+        else:
+            for i in range(len(x)):  # Iterate over each element in the input signal
+                x_upsampled += [x[i]] + list(np.zeros(n-1, dtype=type(x[0])))  # Add the current element and n zeros after each element
+        return x_upsampled
+
+    elif dim == 2:  # If the input signal is 2D
+        if axis == 0:  # If axis input is 0, interpolate over rows
```

Comment:
lets not have them do column and rows, option, if the want to do just a collumn or row, have them call this as a 1 dimensional function, if they pass in a 2->d function lets do nearest neighbor if just upsample, or bi-liniear if interpolating

Reply:
just taking out everything thats not 1-D